File: 232.implement-queue-using-stacks.java
/*
 * @lc app=leetcode id=232 lang=java
 *
 * [232] Implement Queue using Stacks
 */

// @lc code=start

import java.util.Stack;

class MyQueue {

    private Stack<Integer> stack1;
    private Stack<Integer> stack2;

    public MyQueue() {
        this.stack1 = new Stack<>();
        this.stack2 = new Stack<>();
    }

    public void push(int x) {
        this.stack1.push(x);
    }

    public int pop() {
        if (this.stack2.empty()) {
            while (!this.stack1.empty()) {
                this.stack2.push(this.stack1.pop());
            }
        }
        if (this.stack2.empty()) {
            return -1;
        }
        return this.stack2.pop();
    }

    public int peek() {
        if (this.stack2.empty()) {
            while (!this.stack1.empty()) {
                this.stack2.push(this.stack1.pop());
            }
        }
        if (this.stack2.empty()) {
            return -1;
        }
        return this.stack2.peek();
    }

    public boolean empty() {
        return this.stack1.empty() && this.stack2.empty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */
// @lc code=end

File: 232.implement-queue-using-stacks.cpp
/*
 * @lc app=leetcode id=232 lang=cpp
 *
 * [232] Implement Queue using Stacks
 */

// @lc code=start

#include <stack>

using namespace std;

class MyQueue {
private:
    stack<int> s1_;
    stack<int> s2_;

public:
    MyQueue() {
    }

    void push(int x) {
        s1_.push(x);
    }

    int pop() {
        if (s2_.empty()) {
            while (!s1_.empty()) {
                int x = s1_.top();
                s2_.push(x);
                s1_.pop();
            }
        }
        int ret = s2_.top();
        s2_.pop();
        return ret;
    }

    int peek() {
        if (s2_.empty()) {
            while (!s1_.empty()) {
                int x = s1_.top();
                s2_.push(x);
                s1_.pop();
            }
        }
        return s2_.top();
    }

    bool empty() {
        return s1_.empty() && s2_.empty();
    }

};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */
// @lc code=end
